### ISSUE TITLE
Refactor retrieveParticipants() API

### ISSUE THREAD

**Issue by anthonypetersen** [episphere/connect#1176]
`getParticipants` API is relying on `offset` functionality from Firestore as a means of implementing pagination and is causing our usage to be much higher than necessary

Firestore offers `startAfter` function as a way of implementing pagination, though this will require incoming calls to pass in a document ID so our backend knows where they left off. This could be easily returned to a client in a response object if they aren't on the last page.

Potential drawbacks of this:
1) No way to jump to a specific page as we wouldn't be using page numbers anymore, but document IDs, to instruct our query where to look.
2) Calls will need to be made synchronously as each one relies on the previous call in order to get the `startAfter` document ID. This also means if one call fails, the rest of the calls won't happen. This could be resolved by clients implementing a system to try a call again.

**Comment by we-ai** [comment]
Not sure exactly how this API is used.
If slight data lag doesn't matter much, BQ can be used for (quick and inexpensive) paginations.

**Comment by JoeArmani** [comment]
Sounds like a great approach to me.
Firestore Cursors: https://firebase.google.com/docs/firestore/query-data/query-cursors

**Pagination and SMDB:**
This is also used in SMDB: Main menu -> Participants -> dropdown options such as 'verified', 'not yet verified', 'all', etc.

I _think_ we can use Firestore's `.count()` feature, which costs 1 read per 1000 documents counted, to maintain pagination options in SMDB if we want to keep that.

I don't know how those are used in SMDB, so it would be good to discuss those use cases. Those calls may be doing more than necessary in addition to the costly `offset` operation.

**Potentially costly offset operations** (Warren flagged these previously in the codebase, I'm just adding them here for visibility)**:**
`retrieveParticipants` - primary
`retrieveRefusalWithdrawalParticipants`
`retrieveParticipantsEligibleForIncentives`
`getBoxesPagination`

**Comment by anthonypetersen** [comment]
`retrieveParticipants` offset reads account for about 97% of our reads, the others don't really seem to be much of an issue at the moment.

For the purpose of viewing on the SMDB, I feel like full pagination isn't totally necessary and returning 1000 records max would be fine. No one using that dashboard will truly go through pages of records to see thousands of documents, they would use a search functionality that is more precise.

**Comment by anthonypetersen** [comment]
SMDB changes have been moved to dev

**Comment by anthonypetersen** [comment]
@brotzmanmj yes - the dashboard issues you were seeing should be resolved now, although I don't see how that would've been an issue in stage & production, the API code hasn't been touched there

**Comment by brotzmanmj** [comment]
@anthonypetersen Interesting. We were just looking at a filtered list in prod and tried to go to the next page and it kicked us out. Perhaps a different issue?

**Comment by brotzmanmj** [comment]
@erincschwartz do you think you could demo for Tony the issue you were having?

**Comment by anthonypetersen** [comment]
@brotzmanmj I was able to re-create what you're describing, unfortunately I don't think the code I'm adding will fix this. The issue you're seeing has likely been around for quite some time (paginating via participant lookup workflow)

**Comment by brotzmanmj** [comment]
@anthonypetersen Thanks for identifying that. We will gather information from SMDB users as to how much of an issue this is for them in their daily work. And then most likely I would suggest we put this on the list for SMDB updates that are planned for 2025.
